File: macos/AppLogger.swift
import Foundation
import OSLog

class AppLogger {
    static let shared = AppLogger()

    private let appGroupIdentifier = "group.com.lightning.manna"
    let logDir: URL?
    private let queue = DispatchQueue(label: "com.lightning.manna.logger", qos: .utility)
    private let logger = Logger(subsystem: "com.lightning.manna", category: "AppLogger")

    private init() {
        if let containerURL = FileManager.default.containerURL(
            forSecurityApplicationGroupIdentifier: appGroupIdentifier
        ) {
            let logDirURL = containerURL.appendingPathComponent("logs")
            if (try? FileManager.default.createDirectory(at: logDirURL, withIntermediateDirectories: true)) != nil {
                logDir = logDirURL
                return
            }
        }
        logDir = nil

    }

    func logD(_ message: String, tag: String? = nil) { log(message: message, level: LogLevel.debug, tag: tag) }
    func logI(_ message: String, tag: String? = nil) { log(message: message, level: LogLevel.info, tag: tag) }
    func logW(_ message: String, tag: String? = nil) { log(message: message, level: LogLevel.warning, tag: tag) }
    func logE(_ message: String, tag: String? = nil) { log(message: message, level: LogLevel.error, tag: tag) }
    func logF(_ message: String, tag: String? = nil) { log(message: message, level: LogLevel.fatal, tag: tag) }

    private func log(message: String, level: LogLevel, tag: String?) {
        guard let fileURL = logDir?.appendingPathComponent("ios.jsonl") else { return }

        let stackTrace =
            (level == .error || level == .fatal) ? Thread.callStackSymbols.dropFirst().joined(separator: "\n") : nil
        queue.async {
            var entry: [String: Any] = [
                "ts": ISO8601DateFormatter().string(from: Date()),
                "lvl": level.rawValue,
                "msg": message,
                "tag": tag ?? "",
            ]

            if let stack = stackTrace {
                entry["stack"] = stack
            }

            if level != .debug {
                do {
                    let data = try JSONSerialization.data(withJSONObject: entry)
                    if let line = String(data: data, encoding: .utf8) {
                        try line.appendToURL(fileURL)
                    }
                } catch {
                    os_log("Failed to write log: %@", log: .default, type: .error, error.localizedDescription)
                }
            }
        }

        let tagStr = tag ?? "no-tag"
        if let stack = stackTrace {
            logger.log(
                level: level.osLogType,
                "\(level.rawValue, privacy: .public) | \(tagStr, privacy: .public) | \(message, privacy: .public)"
            )
            logger.error("Stack:\n\(stack, privacy: .public)")
        } else {
            logger.log(
                level: level.osLogType,
                "\(level.rawValue, privacy: .public) | \(tagStr, privacy: .public) | \(message, privacy: .public)"
            )
        }

    }
}

enum LogLevel: String {
    case debug = "D"
    case info = "I"
    case warning = "W"
    case error = "E"
    case fatal = "F"

    var osLogType: OSLogType {
        switch self {
        case .debug: return .debug
        case .info: return .info
        case .warning: return .info
        case .error: return .error
        case .fatal: return .fault
        }
    }
}

extension String {
    fileprivate func appendToURL(_ url: URL) throws {
        let line = self + "\n"
        guard let data = line.data(using: .utf8) else { return }

        if FileManager.default.fileExists(atPath: url.path) {
            let fileHandle = try FileHandle(forWritingTo: url)
            try fileHandle.seekToEnd()
            fileHandle.write(data)
            try fileHandle.close()
        } else {
            try data.write(to: url, options: .atomic)
        }
    }
}

File: android/app/src/main/kotlin/com/lightning/manna/AppLogger.kt
package com.lightning.manna

import android.content.Context
import android.util.Log
import org.json.JSONObject
import java.io.File
import java.io.FileWriter
import java.time.Clock
import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneOffset


object AppLogger {
    private const val APP_TAG = "Manna"
    lateinit var logDir: File

    fun init(context: Context) {
        logDir = File(context.filesDir, "logs")
        if (!logDir.exists()) logDir.mkdirs()
    }

    fun logD(message: String, tag: String? = null) = log(LogType.DEBUG, message, tag)
    fun logI(message: String, tag: String? = null) = log(LogType.INFO, message, tag)
    fun logW(message: String, tag: String? = null) = log(LogType.WARNING, message, tag)
    fun logE(message: String, tag: String? = null) = log(LogType.ERROR, message, tag)
    fun logF(message: String, tag: String? = null) = log(LogType.FATAL, message, tag)

    private fun log(
        level: LogType,
        message: String,
        tag: String?,
    ) {
        val entry = JSONObject().apply {
            put("ts", OffsetDateTime.now(ZoneOffset.UTC).toString())
            put("lvl", level)
            put("msg", message)
            put("tag", tag ?: "")
        }

        val stackTraceString = Thread.currentThread().stackTrace
            .drop(5)
            .joinToString("\n") { it.toString() }
        if (level == LogType.ERROR || level == LogType.FATAL) {
            entry.put("stack", stackTraceString)
        }

        try {
            val logFile = File(logDir, "android.jsonl")
            FileWriter(logFile, true).use { writer ->
                writer.append(entry.toString())
                writer.append("\n")
            }
        } catch (e: Exception) {
            Log.e(APP_TAG, "Failed to write log to file", e)
        }

        val logTag = tag ?: ""
        val fullMessage = "$level | $logTag | $message"

        when (level) {
            LogType.DEBUG -> Log.d(APP_TAG, fullMessage)
            LogType.INFO -> Log.i(APP_TAG, fullMessage)
            LogType.WARNING -> Log.w(APP_TAG, fullMessage)
            LogType.ERROR, LogType.FATAL -> {
                Log.e(APP_TAG, fullMessage)
                Log.e(APP_TAG, "Stack Trace:\n${stackTraceString}")
            }
        }
    }
}

enum class LogType {
    DEBUG, INFO, WARNING, ERROR, FATAL
}